transforms/p2_custom_models.py

```diff
@@ -0,0 +1,53 @@
+"""Add keyvalues to Portal 2 test element entities to automatically handle custom models."""
+
+from srctools import conv_int
+
+from hammeraddons.bsp_transform import trans, Context
+
+SUPPORTED_ENTS = [
+    "prop_weighted_cube",
+    "prop_button",
+    "prop_floor_cube_button",
+    "prop_floor_ball_button",
+    "prop_under_button",
+    "prop_under_floor_button",
+    "prop_wall_projector",
+    "prop_glados_core",
+    # TODO: needs extra VScript for the sprite, otherwise should work
+    # "npc_security_camera",
+    "npc_rocket_turret",
+    "npc_personality_core",
+    "prop_exploding_futbol",
+    "prop_glass_futbol",
+
+    # Not supported:
+    # prop_testchamber_door - reverses animations
+    # prop_tractor_beam - breaks animations
+    # prop_linked_portal_door - breaks animations
+    # prop_monster_box - swaps models dynamically
+]
+
+@trans('Portal 2 Custom Models')
+def p2_custom_models(ctx: Context) -> None:
+    """Add keyvalues to Portal 2 test element entities to automatically handle custom models."""
+    for ent in ctx.vmf.entities:
+        if ent['classname'] not in SUPPORTED_ENTS:
```

Comment:
It'd be more efficient here to loop over `SUPPORTED_ENTS`, then use `ctx.vmf.by_class[classname]`. The VMF class keeps track of classnames, so it can directly give you those entities. If not present it'll just give an empty set.

Reply:
Oh, you mean like it pre-builds the list of entities for each classname always instead of only when you ask for it?

Reply:
I did have it using that before but changed it when I added the multiple entities because I assumed that would make it loop over the entire list of entities in the map each time I ask for a specific classname

Reply:
When the `classname` keyvalue on entities are modified it updates the `by_class` dict yeah.